docs/Middleware-Guide.md

```diff
@@ -0,0 +1,82 @@
+
+## Integration with EigenLayer Contracts:
+In this section, we will explain various API interfaces that EigenLayer provides which are essential for middlewares to integrate with EigenLayer. 
+
+### *Opting into middleware*
+In order for any EigenLayer operator to be able to opt-in to a middleware, EigenLayer provides two interfaces: `optIntoSlashing(..)` and `recordFirstStakeUpdate(..)`. The sequential flow for opting into a middleware using these functions is as follows:
+1. The operator first opts into slashing by calling  `optIntoSlashing(..)` in `Slasher.sol`, where it has to specify the address of the middleware's ServiceManager contract in the argument. This step results in the operator giving permission to the middleware's ServiceManager contract to slash the operator via EigenLayer, if the operator is ever proven to have engaged in adversarial behavior while responding to the middleware's task. Successful call to  `optIntoSlashing(..)` emits the `OptedIntoSlashing(..)` event.
+2. Next, the operator needs to register with the middleware on chain via a middleware-specific registry contract (see the *Guide To Provided Middleware Contracts* section for examples). To integrate with EigenLayer, the middleware's Registry contract provides a registration endpoint that calls on the `recordFirstStakeUpdate(..)` in the middleware's ServiceManager contract which in turn calls the `recordFirstStakeUpdate(..)` function in `Slasher.sol`. On successful execution of this function call, the event `MiddlewareTimesAdded(..)` is emitted and the operator has to start serving the tasks from the middleware.
+
+The following figure illustrates the above flow: 
+![Operator opting-in](./images/operator_opting.png)
+
+
+### *Deregistering from middleware*
+In order for any EigenLayer operator to be able to de-register from a middleware, EigenLayer provides the interface `recordLastStakeUpdateAndRevokeSlashingAbility(..)`. Essentially, in order for an operator to de-register from a middleware, the operator has to call `recordLastStakeUpdateAndRevokeSlashingAbility(..)` in EigenLayer's `Slasher.sol` via the middleware's ServiceManager contract. One important thing to note is that in order to call this function, the information on the latest block number until which the operator is required to serve the middleware needs to be passed.
+
+The following figure illustrates the above flow in which the operator calls on a `deregister(..)` function in a sample Registry contract.
+![Operator deregisteringn](./images/operator_deregister.png)
+
+### *Staker opting-in with a middleware*
+In order for any staker to participate in responding to the tasks from the middleware without running the middleware's off-chain software container themself, the staker can pursue delegation. The sequential flow for delegation process is as follows:
```

Comment:
why is this section in this document at all? should AVS developers care about this?

Reply:
hmm good point. can we lean-in to the AVS developer perspective here?
we should probably explain somewhere that EigenLayer has a bit of a pull-push model with AVSs, where AVSs "pull" stake updates (or the info for them, at least) and "push" a notification to EigenLayer when they perform these updates. I think something brief is OK for now; we can link maybe to the detailed documentation about withdrawals if people want to understand more?